grpc/server.go

```diff
@@ -68,3 +68,58 @@ func (s *Server) List(ctx context.Context, req *v1.ListRequest) (*v1.ListRespons
 
 	return repoList.ToProto(), nil
 }
+
+// gRPCChunkSender is a zoekt.Sender that sends small chunks of FileMatches to the provided gRPC stream.
+func gRPCChunkSender(ss v1.WebserverService_StreamSearchServer) zoekt.Sender {
+	f := func(r *zoekt.SearchResult) {
+		result := r.ToProto()
+
+		if len(result.GetFiles()) == 0 { // stats-only result, send it immediately
+			_ = ss.Send(result)
+			return
+		}
+
+		// Otherwise, chunk the file matches into multiple responses
+
+		statsSent := false
+		numFilesSent := 0
+
+		sendFunc := func(filesChunk []*v1.FileMatch) error {
+			numFilesSent += len(filesChunk)
+
+			var stats *v1.Stats
+			if !statsSent { // We only send stats back on the first chunk
+				statsSent = true
+				stats = result.GetStats()
+			}
```

Comment:
why the first chunk and not the last chunk? It seems like it could be problematic to update the stats before we get the results represented by the stats.

Reply:
I asked for this change. We don't make decisions on stats anywhere, we only use it for observability. I thought this woudl be more useful for the case where we hit a limit in frontend and call cancel, then miss out on the stats.

Reply:
Ah, got it, makes sense to me 👍